src/iam/iam-base.ts

```diff
@@ -199,98 +195,127 @@ export class IAMBase {
               }
             ]
-          });
-          if (accounts.length < 1) {
-            await metamaskProvider.request({
-              method: "wallet_requestPermissions",
-              params: [
-                {
-                  eth_accounts: {}
-                }
-              ]
-            });
-          }
-        }
-        const provider = new providers.Web3Provider(metamaskProvider);
-        this._signer = provider.getSigner();
-        const publicKey = await this.getPublicKey();
-        this._didSigner = new IdentityOwner(this._signer, this._provider, publicKey);
-        return;
+          };
+      const accounts: string[] = await metamaskProvider.request(requestObject);
+
+      if (!initializeMetamask && accounts.length < 1) {
+        await metamaskProvider.request({
+          method: "wallet_requestPermissions",
+          params: [
+            {
+              eth_accounts: {}
+            }
+          ]
+        });
       }
-      this._transactionOverrides = {
-        gasLimit: hexlify(4900000),
-        gasPrice: hexlify(0.1)
-      };
-      this._walletConnectProvider = new WalletConnectProvider({
-        rpc: {
-          [this._connectionOptions.chainId]: this._connectionOptions.rpcUrl
-        },
-        infuraId: this._connectionOptions.infuraId,
-        bridge: this._connectionOptions.bridgeUrl
-      });
-      await this._walletConnectProvider.enable();
-      this._signer = new providers.Web3Provider(this._walletConnectProvider).getSigner();
-      const publicKey = await this.getPublicKey();
-      this._didSigner = new IdentityOwner(this._signer, this._provider, publicKey);
+      const provider = new providers.Web3Provider(metamaskProvider);
+      this._signer = provider.getSigner();
+      return;
     }
+    this._transactionOverrides = {
+      gasLimit: hexlify(4900000),
+      gasPrice: hexlify(0.1)
+    };
+    this._walletConnectProvider = new WalletConnectProvider({
+      rpc: {
+        [this._connectionOptions.chainId]: this._connectionOptions.rpcUrl
+      },
+      infuraId: this._connectionOptions.infuraId,
+      bridge: this._connectionOptions.bridgeUrl
+    });
+    await this._walletConnectProvider.enable();
+    this._signer = new providers.Web3Provider(this._walletConnectProvider).getSigner();
   }
 
-  private async getPublicKey() {
+  private async generateIdentityProof() {
     if (this._signer) {
+      const header = {
+        alg: "ES256",
+        typ: "JWT"
+      };
+      const encodedHeader = base64url(JSON.stringify(header));
       const address = await this._signer.getAddress();
-      const hash = keccak256(address);
+      const payload = {
+        iss: `did:ethr:${address}`,
+        claimData: {
+          blockNumber: await this._provider?.getBlockNumber()
+        }
+      };
+
+      const encodedPayload = base64url(JSON.stringify(payload));
+      return { encodedHeader, encodedPayload };
+    }
+    throw Error("Signer not initialized");
```

Comment:
I would use the same `ERROR_MESSAGES.SIGNER_NOT_INITIALIZED` here

Reply:
👍